vshard/router/init.lua

```diff
@@ -945,6 +1051,213 @@ local function router_map_callrw(router, func, args, opts)
+        futures = table_new(0, #grouped_buckets)
+        for uuid, buckets in pairs(grouped_buckets) do
+            replicaset = router.replicasets[uuid]
+            call_args = {'storage_ref_with_lookup', rid, timeout, buckets}
+            res, err = replicaset:callrw('vshard.storage._call', call_args, call_opts)
+            if res == nil then
+                err_uuid = uuid
+                goto fail
+            end
+            futures[uuid] = res
+            rs_count = rs_count + 1
+        end
+
+        -- We hope that router's cache is up to date and only a single
+        -- loop iteration is required.
+        if not preallocated then
+            replicasets = table_new(0, rs_count)
+            preallocated = true
+        end
+
+        -- Wait for the refs to be done and collect moved buckets.
+        bucket_ids = {}
+        for uuid, future in pairs(futures) do
+            res, err = future_wait(future, timeout)
+            -- Handle netbox error first.
+            if res == nil then
+                err_uuid = uuid
+                goto fail
+            end
+            -- Ref returns nil,err or moved buckets.
+            res, err = res[1], res[2]
+            if res == nil then
+                err_uuid = uuid
+                goto fail
+            end
+            for _, bucket_id in pairs(res) do
+                bucket_reset(router, bucket_id)
+                table.insert(bucket_ids, bucket_id)
+            end
+            replicasets[uuid] = router.replicasets[uuid]
+            timeout = deadline - fiber_clock()
+        end
+    end
+
+    -- Map stage.
+    map = table_new(0, rs_count)
+    futures = table_new(0, rs_count)
+    args = {'storage_map', rid, func, args}
+    -- Send map requests.
+    for uuid, rs in pairs(replicasets) do
+        res, err = rs:callrw('vshard.storage._call', args, call_opts)
+        if res == nil then
+            err_uuid = uuid
+            goto fail
+        end
+        futures[uuid] = res
+    end
+
+    -- Reduce stage.
+    -- Collect map responses (refs were deleted by the storages for non-error results).
+    for uuid, future in pairs(futures) do
+        res, err = future_wait(future, timeout)
+        if res == nil then
+            err_uuid = uuid
+            goto fail
+        end
+        -- Map returns true,res or nil,err.
+        ok, res = res[1], res[2]
+        if ok == nil then
+            err = res
+            err_uuid = uuid
+            goto fail
+        end
+        if res ~= nil then
+            -- Execute callback on the replicaset result.
+            if callback ~= nil then
+                ok, cb_res = pcall(callback, uuid, res, cb_res)
+                if not ok then
+                    err_uuid = uuid
+                    err = lerror.make(cb_res)
+                    goto fail
+                end
+                map[uuid] = cb_res
+            else
+                map[uuid] = res
```

Comment:
Should I rather return `{res}` instead, as `map_callrw` does? It has a strange comment:
```
-- Store as a table so in future it could be extended for
-- multireturn.
```
What is it all about?